pkg/genericapiserver/serve.go

```diff
@@ -0,0 +1,256 @@
+
+			// listen again, but never retry instead of fail.
+			func() {
+				lock.Lock()
+				defer lock.Unlock()
+				for {
+					time.Sleep(15 * time.Second)
+
+					ln, err = net.Listen("tcp", addr)
+					if err == nil {
+						return
+					}
+					select {
+					case <-stopCh:
+						return
+					default:
+					}
+					glog.Errorf("Error listening on %v (%v); will try again.", addr, err)
+				}
+			}()
+
+			select {
+			case <-stopCh:
+				return
+			default:
+			}
+		}
+	}()
+
+	return tcpAddr.Port, nil
+}
+
+// getNamedCertificateMap returns a map of strings to *tls.Certificate, suitable for use in
+// tls.Config#NamedCertificates. Returns an error if any of the certs cannot be loaded.
+// Returns nil if len(namedKeyCerts) == 0
+func getNamedCertificateMap(namedCertKeys []NamedCertKey) (map[string]*tls.Certificate, error) {
+	if len(namedCertKeys) == 0 {
+		return nil, nil
+	}
+
+	// load keys
+	tlsCerts := make([]tls.Certificate, len(namedCertKeys))
+	for i := range namedCertKeys {
+		var err error
+		nkc := &namedCertKeys[i]
+		tlsCerts[i], err = tls.LoadX509KeyPair(nkc.CertFile, nkc.KeyFile)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// register certs with implicit names first, reverse order such that earlier trump over the later
+	tlsCertsByName := map[string]*tls.Certificate{}
+	for i := len(namedCertKeys) - 1; i >= 0; i-- {
+		nkc := &namedCertKeys[i]
+		if len(nkc.Names) > 0 {
+			continue
+		}
+		cert := &tlsCerts[i]
+
+		// read names from certificate common names and DNS names
+		if len(cert.Certificate) == 0 {
+			return nil, fmt.Errorf("no certificate found in %q", nkc.CertFile)
+		}
+		x509Cert, err := x509.ParseCertificate(cert.Certificate[0])
+		if err != nil {
+			return nil, fmt.Errorf("parse error for certificate in %q: %v", nkc.CertFile, err)
+		}
+		if len(x509Cert.Subject.CommonName) > 0 {
+			tlsCertsByName[x509Cert.Subject.CommonName] = cert
+		}
+		for _, san := range x509Cert.DNSNames {
+			tlsCertsByName[san] = cert
+		}
+		// intentionally all IPs in the cert are ignored as SNI forbids passing IPs
+		// to select a cert. Before go 1.6 the tls happily passed IPs as SNI values.
+	}
+
+	// register certs with explicit names last, overwriting every of the implicit ones,
+	// again in reverse order.
+	for i := len(namedCertKeys) - 1; i >= 0; i-- {
+		nkc := &namedCertKeys[i]
+		if len(nkc.Names) == 0 {
+			continue
+		}
+		for _, name := range nkc.Names {
+			tlsCertsByName[name] = &tlsCerts[i]
+		}
+	}
+
+	return tlsCertsByName, nil
+}
+
+// tcpKeepAliveListener sets TCP keep-alive timeouts on accepted
+// connections. It's used by ListenAndServe and ListenAndServeTLS so
+// dead TCP connections (e.g. closing laptop mid-download) eventually
+// go away.
+type tcpKeepAliveListener struct {
+	*net.TCPListener
+}
+
+func (ln tcpKeepAliveListener) Accept() (c net.Conn, err error) {
+	tc, err := ln.AcceptTCP()
+	if err != nil {
+		return
+	}
+	tc.SetKeepAlive(true)
+	tc.SetKeepAlivePeriod(3 * time.Minute)
```

Comment:
Is it ok to hard-code this value? Maybe at least make it a constant if it's not configurable?